### PR TITLE
Don't collect preprocssor flags separately from compiler flags

### LINKED ISSUE #4272: Meson does not properly uses CXXFLAGS in compiler test
Meson should use CXXFLAGS env variable for compiler tests, it does for some but not for others.

Here the `has_library` test (remove some unnecessary info):

```
Running compile:
Command line:  c++ -I/Users/epirat/Projects/Misc/plprefix/include -L/Users/epirat/Projects/Misc/plprefix/lib /var/folders/rw/l97gj1cd6vd_wrf_zjcbycjc0000gn/T/tmpnho14245/testfile.cpp -pipe -o /var/folders/rw/l97gj1cd6vd_wrf_zjcbycjc0000gn/T/tmpnho14245/output.exe -O0 -fpermissive -lSPVRemapper -Wl,-undefined,dynamic_lookup 
```

As can be seen, meson properly uses the `CXXFLAGS` `-I/Users/epirat/Projects/Misc/plprefix/include`.

Here the `get_define`:

```
Running compile:
Command line:  c++ /var/folders/rw/l97gj1cd6vd_wrf_zjcbycjc0000gn/T/tmph8m3486_/testfile.cpp -pipe -E -P -O0 -fpermissive 
```

As can be seen, it lacks the `CXXFLAGS`.

This is very inconsistent, and leads to weird bugs, even worse than before as now it finds the library but fails with a (for the user) inexplicable other error because headers of that library are not properly found.

### PR DESCRIPTION
**Pull request by Ericson2314**
I recall that @jpakkane never wanted this, but @nirbheek did, but then @nirbheek changed his mind.

I am fine either way except for the cross inconsistency that exists today: There is no `c_preproc_args` or similar one can put in the cross file, so no way to replicate the effect of `CPPFLAGS` during cross compilation.

Closes #4272

CC @dcbaker

### CODE REVIEW

**Comment by Ericson2314** [comment]
https://github.com/mesonbuild/meson/pull/5123/files#diff-2736199afea34a4f93305938406f439a the `run_unittest.py` changes nicely summarize the change in semantics. (`LDFLAGS` wasn't read before either.)

File: mesonbuild/compilers/c.py [hunk 1]
```diff
@@ -424,11 +424,8 @@ def _get_compiler_check_args(self, env, extra_args, dependencies, mode='compile'
             for_machine = MachineChoice.BUILD
         else:
             for_machine = MachineChoice.HOST
-        if mode == 'preprocess':
-            # Add CPPFLAGS from the env.
-            args += env.coredata.get_external_preprocess_args(for_machine, self.language)
-        elif mode == 'compile':
-            # Add CFLAGS/CXXFLAGS/OBJCFLAGS/OBJCXXFLAGS from the env
+        if mode == 'compile' or mode == 'preprocess':
```
Comment (dcbaker) [inline mesonbuild/compilers/c.py]:
```python
if mode in {'compile', 'preprocess'}:
```

**Comment by jpakkane** [comment]
This is a behavioural change (or at the very least the old way was not well documented), needs a release note snippet. And @nirbheek is this change ok to you? Can we still do the MSVC asm thing that (if I remember correctly) was the main reason to keep `CPPFLAGS` as a separate entity?

**Comment by Ericson2314** [comment]
https://github.com/mesonbuild/meson/issues/4272#issuecomment-424952670 ah here is where @nirbheek mentioned changing his mind, but we should get confirmation because that doesn't mention the asssembly / preprocessor run separately side of things. @nirbheek please chime in!

**Comment by Ericson2314** [comment]
Chasing that I got to #3916, where @nirbheek mentions the assembly issue. If this is merged I believe nothing more from that PR is needed? (Somebody commented there that #3894 already did the rest.)

**Comment by Ericson2314** [comment]
@nirbheek approved the concept on IRC (yay!), and also wants to give the code a once-over. Thanks for both!

**Pull request merged**